File: Java/LinkedListPractice/src/DoublyLinkedList/DoublyLinkedList.java
package DoublyLinkedList;

public class DoublyLinkedList {

    private DLLNode head;

    private DLLNode tail;

    private int length;

    public DoublyLinkedList() {
        this.length = 0;
        this.head = null;
        this.tail = null;
    }

    public void append(int data) {
        DLLNode newNode = new DLLNode(data);
        if (length == 0) {
            this.head = newNode;
        } else {
            newNode.previous = this.tail;
            this.tail.next = newNode;
        }
        this.tail = newNode;
        this.length++;
    }

    public boolean contains(int data) {
        DLLNode currentNode = this.head;
        while (currentNode != null) {
            if (currentNode.data == data) {
                return true;
            }
            currentNode = currentNode.next;
        }
        return false;
    }

    public Object get(int index) {
        int currentIndex = 0;
        DLLNode currentNode = this.head;
        while (currentNode != null) {
            if (index == currentIndex) {
                return currentNode.data;
            }
            currentNode = currentNode.next;
            currentIndex++;
        }
        return null;
    }

    public String getDebugInfo(int index) {
        int currentIndex = 0;
        DLLNode currentNode = this.head;
        while (currentNode != null) {
            if (index == currentIndex) {
                return "NODE AT INDEX " + index + ": [" + currentNode.toString() + "]";
            }
            currentNode = currentNode.next;
            currentIndex++;
        }
        return "NODE AT INDEX " + index + ": [ERROR: NO NODE FOUND]";
    }

    public String getHead() {
        if (this.head == null) {
            return "Head is null";
        } else {
            return "HEAD: [" + this.head.toString() + "]";
        }
    }

    public int getLength() {
        return length;
    }

    public String getTail() {
        if (this.tail == null) {
            return "Tail is null";
        } else {
            return "TAIL: ["  + this.tail.toString() + "]";
        }
    }

    public void insertAt(int index, int data) {
        if (index > this.length) {
            return;
        } else if (index == this.length) {
            this.append(data);
        } else if (index == 0) {
            this.prepend(data);
        } else {
            DLLNode currentNode = this.head;
            int currentIndex =  0;
            while (currentIndex < index
                    && currentIndex< this.length
                    && currentNode.next != null
            ) {
                System.out.println(currentNode.toString());
                currentNode = currentNode.next;
                currentIndex++;
            }
            DLLNode newNode = new DLLNode(data);
            if (currentNode.previous != null) {
                newNode.previous = currentNode.previous;
                currentNode.previous.next = newNode;
                currentNode.previous = newNode;
                newNode.next = currentNode;
                this.length++;
            }
        }
    }

    public void prepend(int data) {
        DLLNode newNode = new DLLNode(data);
        if (length == 0) {
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head.previous = newNode;
        }
        this.head = newNode;
        this.length++;
    }

    public void printAll() {
        DLLNode currentNode = this.head;
        while (currentNode != null) {
            if (currentNode == this.head) {
                System.out.println("***************Head****************");
            }
            System.out.println(currentNode.data);
            if (currentNode == this.tail) {
                System.out.println("***************Tail****************");
            }
            currentNode = currentNode.next;
        }
    }

    public void printAllReverse() {
        DLLNode currentNode = this.tail;
        while (currentNode != null) {
            if (currentNode == this.tail) {
                System.out.println("***************Tail****************");
            }
            System.out.println(currentNode.data);
            if (currentNode == this.head) {
                System.out.println("***************Head****************");
            }
            currentNode = currentNode.previous;
        }
    }

    public void removeFrom(int index) {
        if (index >= this.length || index < 0 || this.head == null || this.tail == null) {
            return;
        } else if (index == 0) {
            if(this.head.next != null) {
                this.head.next.previous = null;
                this.head = this.head.next;
            } else {
                this.head = null;
            }
            this.length--;
        } else if (index == this.length - 1) {
            if(this.tail.previous != null) {
                this.tail.previous.next = null;
                this.tail = this.tail.previous;
                this.length--;
            }
        } else {
            DLLNode currentNode = this.head;
            int currentIndex =  0;
            while (currentIndex < index && currentNode.next != null
            ) {
                currentNode = currentNode.next;
                currentIndex++;
            }
            if (currentNode.previous != null) {
                currentNode.previous.next = currentNode.next;
                assert currentNode.next != null;
                currentNode.next.previous = currentNode.previous;
                this.length--;
            }
        }
    }

}

File: Kotlin/LinkedListPractice/src/doublyLinkedList/DoublyLinkedList.kt
package doublyLinkedList

class DoublyLinkedList(
    var head: DLLNode?,
    var tail: DLLNode?,
    var size: Int = 0
) {

    fun append(data: Int) {
        val newNode = DLLNode(data=data, next=null, previous=null)
        if (size == 0) {
            head = newNode
        } else {
            newNode.previous = tail
            tail?.next = newNode;
        }
        tail = newNode
        size++
    }

    fun contains(data: Int): Boolean {
        var currentNode = head
        while (currentNode != null) {
            if (currentNode.data == data) {
                return true
            }
            currentNode = currentNode.next
        }
        return false
    }

    fun get(index: Int): Int? {
        if (index >= 0 && index <= size -1) {
            var currentIndex = 0
            var currentNode = head
            while (currentIndex < index && currentNode != null) {
                currentNode = currentNode.next
                currentIndex++
            }
            if (currentIndex == index && currentNode !=null) {
                return currentNode.data
            }
        }
        return null
    }

    fun insertAt(index: Int, data: Int) {
        if (index < 0 ||  index > size) {
            return
        } else if (index == size){
            this.append(data)
        } else if (index == 0) {
            this.prepend(data)
        } else {
            var currentNode = head
            var currentIndex = 0
            while (currentIndex < index && currentNode != null) {
                currentNode = currentNode.next
                currentIndex++
            }
            if (currentIndex == index && currentNode != null) {
                val newNode = DLLNode(data=data, next=currentNode, previous=currentNode.previous)
                currentNode.previous?.next = newNode
                currentNode.previous = newNode
                size++
            }
        }
    }

    fun prepend(data: Int) {
        val newNode = DLLNode(data=data, next=null, previous=null)
        if (size == 0) {
            tail = newNode
        } else {
            newNode.next = head
            head?.previous = newNode
        }
        head = newNode
        size++
    }

    fun printAll() {
        println("************LinkedList***************")
        if (size > 0) {
            var currentNode = head
            while(currentNode != null) {
                if (currentNode == head) {
                    println("*-*-*-*-*-Head*-*-*-*-*-*-*-")
                }
                println(currentNode.data)
                if (currentNode == tail) {
                    println("*-*-*-*-*-Tail*-*-*-*-*-*-*-")
                }
                currentNode = currentNode.next
            }
        } else {
            println("The Linked List is currently empty")
        }
        println("************LinkedList***************")
    }

    fun removeFrom(index: Int) {
        if (index < 0 || index > size - 1) {
            return
        } else if (index == 0) {
            head = head?.next
            head?.previous = null
            size--
        } else if (index == size -1) {
            tail?.previous?.next = null
            tail = tail?.previous
            size--
        } else {
            var currentIndex = 0
            var currentNode = head
            while (currentIndex < index && currentNode != null) {
                currentIndex++
                currentNode = currentNode.next
            }
            if (currentIndex == index && currentNode != null) {
                currentNode.previous?.next = currentNode.next
                currentNode.next?.previous = currentNode.previous
                size--
            }
        }
    }
}